File: ios/Sources/OberonTUI/TerminalView.swift
// TerminalView — SwiftUI cell-grid view backed by TerminalState.
//
// Scaffold-grade: renders each row as a single Text via an attributed
// string built from the cells. Good enough to see Oberon output on
// screen and exercise the runtime end-to-end. The web port runs span-
// per-cell so per-cell background colour can be styled; the SwiftUI
// equivalent (separate styled Text per cell, or a Canvas drawing) is
// the next iteration once oed itself drives the contents.
//
// Geometry: a monospace font + GeometryReader measures available
// space and reports rows/cols back to TerminalState. The view
// invokes `onReady` once after the first stable resize — the
// embedding app uses that to defer its `<Entry>__init()` until
// after the terminal knows its real dimensions, otherwise paint
// happens at the default 24×80 then gets resized away.
//
// Keyboard input: an invisible KeyboardCapture overlay sits in the
// background as a first responder, mapping UIPress events to Oberon
// codes via oc_dispatch_key. SwiftUI's own .onKeyPress is unreliable
// on iPad — focus state doesn't propagate the way it does on macOS.

import SwiftUI
import UIKit
import OberonRuntime

public struct TerminalView: View {
    @ObservedObject private var state = TerminalState.shared
    @State private var hasReady = false

    /// Monospaced font used for cell measurement and rendering. The
    /// 3270 font is the canonical match for the palette; the host app
    /// should register it (Fonts/UIAppFonts) and pass its name here.
    /// Defaults to the system monospaced face.
    public let fontName: String?
    public let fontSize: CGFloat
    /// Fires once after the first stable resize. The host calls
    /// `<Entry>__init()` (or any Oberon-side bootstrap) here so the
    /// initial paint happens at the real terminal dimensions, not
    /// the 24×80 defaults that get overwritten by GeometryReader's
    /// later resize.
    public let onReady: (() -> Void)?

    public init(fontName: String? = nil,
                fontSize: CGFloat = 16,
                onReady: (() -> Void)? = nil) {
        self.fontName = fontName
        self.fontSize = fontSize
        self.onReady  = onReady
    }

    public var body: some View {
        GeometryReader { proxy in
            let metrics = cellMetrics()
            let cols = max(1, Int(proxy.size.width  / metrics.width))
            let rows = max(1, Int(proxy.size.height / metrics.height))
            ZStack(alignment: .topLeading) {
                // First responder for hardware keyboard events. Has
                // no visible content; lives behind the cell grid.
                KeyboardCapture()
                    .allowsHitTesting(false)

                VStack(alignment: .leading, spacing: 0) {
                    ForEach(0..<state.displayed.count, id: \.self) { r in
                        rowView(r, metrics: metrics)
                            .frame(height: metrics.height, alignment: .leading)
                    }
                    Spacer(minLength: 0)
                }
            }
            .frame(width: proxy.size.width, height: proxy.size.height, alignment: .topLeading)
            .background(Color.black)
            .foregroundColor(Palette.color(4) ?? .green)   // ColorGreen as default
            .font(.system(size: fontSize, design: .monospaced))
            .onAppear {
                state.resize(rows: rows, cols: cols)
                fireReadyIfStable(rows: rows, cols: cols)
            }
            .onChange(of: proxy.size) { _, _ in
                state.resize(rows: rows, cols: cols)
                fireReadyIfStable(rows: rows, cols: cols)
                // Resize doesn't wipe content (the new resize keeps
                // cells in place), but the Oberon side has no idea
                // its grid changed shape — its handleKey-driven
                // Refresh is the only repaint trigger. Dispatch
                // Ctrl-L: oed's handler binds it to "clear status"
                // and unconditionally calls Refresh afterwards, so
                // it re-lays out for the new dimensions without
                // side effects on buffer state.
                if hasReady {
                    oc_dispatch_key(12)
                }
            }
        }
    }

    private func fireReadyIfStable(rows: Int, cols: Int) {
        // Wait until we have realistic dimensions before triggering
        // the host's bootstrap — initial GeometryReader pass can
        // report 0×0 / 1×1 which would have the Oberon side paint
        // a useless layout that gets immediately resized over.
        guard !hasReady, rows > 1, cols > 1 else { return }
        hasReady = true
        onReady?()
    }

    @ViewBuilder
    private func rowView(_ r: Int, metrics: (width: CGFloat, height: CGFloat)) -> some View {
        let row = state.displayed[r]
        ZStack(alignment: .topLeading) {
            // Per-cell styling via AttributedString: each cell becomes
            // a one-character span carrying its fg / bg / bold / italic
            // / underline / reverse-video attributes. Monospaced font
            // keeps widths uniform across bold + italic.
            Text(attributedRow(row))
                .lineLimit(1)
                .truncationMode(.tail)

            // Cursor: solid block over the cell at (curCol, curRow).
            // .blendMode(.difference) inverts the underlying glyph so
            // the character reads through the cursor — classic terminal
            // block-cursor look.
            if state.cursorVisible && r == state.curRow {
                Rectangle()
                    .fill(Palette.color(4) ?? .green)
                    .frame(width: metrics.width, height: metrics.height)
                    .offset(x: CGFloat(state.curCol) * metrics.width)
                    .blendMode(.difference)
            }
        }
    }

    private func attributedRow(_ row: [Cell]) -> AttributedString {
        let defaultFg = Palette.color(4) ?? Color.green   // ColorGreen
        let defaultBg = Color.black
        var out = AttributedString("")
        for cell in row {
            var span = AttributedString(String(cell.ch))
            let cellFg = Palette.color(cell.fg) ?? defaultFg
            let cellBg = Palette.color(cell.bg) ?? defaultBg
            let (fg, bg) = cell.attr.contains(.reverse)
                ? (cellBg, cellFg)
                : (cellFg, cellBg)
            span.foregroundColor = fg
            span.backgroundColor = bg

            var font: Font = .system(size: fontSize, design: .monospaced)
            if cell.attr.contains(.bold)   { font = font.weight(.bold) }
            if cell.attr.contains(.italic) { font = font.italic() }
            span.font = font

            if cell.attr.contains(.underline) {
                span.underlineStyle = .single
            }
            out.append(span)
        }
        return out
    }

    private func cellMetrics() -> (width: CGFloat, height: CGFloat) {
        // Rough metric — UIFont's monospaced-digit width is close
        // enough for layout. The actual SwiftUI Text uses kerning
        // that matches, since we set the same font.
        let font: UIFont = {
            if let name = fontName, let f = UIFont(name: name, size: fontSize) {
                return f
            }
            return UIFont.monospacedSystemFont(ofSize: fontSize, weight: .regular)
        }()
        let attrs: [NSAttributedString.Key: Any] = [.font: font]
        let w = ("M" as NSString).size(withAttributes: attrs).width
        let h = font.lineHeight
        return (w, h)
    }
}

File: android/src/main/kotlin/com/oneav/oberon/tui/TerminalView.kt
// TerminalView — Jetpack Compose cell-grid view backed by
// TerminalState. Scaffold-grade: renders each row as a single
// AnnotatedString whose spans carry the cell's fg / bg / attr; a
// cursor overlay uses BlendMode.Difference to invert the glyph
// underneath. Matches the iOS TerminalView so paint looks the same
// across hosts.
//
// Keyboard: Modifier.onPreviewKeyEvent captures hardware key
// events and dispatches into OberonBridge.ocDispatchKey. We claim
// focus with FocusRequester so the view receives keys immediately
// on first composition. Autorepeat: Android's dispatch driver
// re-fires KeyEventType.KeyDown for held keys automatically at the
// system rate, so we don't run a Timer of our own (unlike iOS).
//
// Geometry: BoxWithConstraints measures available space, we
// compute cellW/cellH from a monospaced font metric, and report
// rows/cols to TerminalState before any Oberon code reads them.
// After the first stable resize we fire `onReady`, and the host
// calls its Oberon `<Entry>__init()` there so the initial paint
// lands at the real dimensions.

package com.oneav.oberon.tui

import androidx.compose.foundation.background
import androidx.compose.foundation.focusable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.text.BasicText
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.input.key.KeyEventType
import androidx.compose.ui.input.key.onPreviewKeyEvent
import androidx.compose.ui.input.key.type
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.rememberTextMeasurer
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlin.math.max

@Composable
fun TerminalView(
    fontSize: Int = 16,
    onReady: (() -> Unit)? = null,
) {
    val displayed by TerminalState.displayed
    val cursorVisible by TerminalState.cursorVisible
    val focusRequester = remember { FocusRequester() }
    var hasReady by remember { mutableStateOf(false) }

    LaunchedEffect(Unit) { focusRequester.requestFocus() }

    BoxWithConstraints(
        modifier = Modifier
            .fillMaxSize()
            .background(Palette.defaultBg)
            .focusRequester(focusRequester)
            .focusable()
            .onPreviewKeyEvent { evt ->
                // Only dispatch on KeyDown; Android re-fires KeyDown
                // for autorepeat so we don't need a Timer.
                if (evt.type != KeyEventType.KeyDown) return@onPreviewKeyEvent false
                val code = mapKeyEvent(evt) ?: return@onPreviewKeyEvent false
                OberonBridge.ocDispatchKey(code)
                true
            }
    ) {
        val density = LocalDensity.current
        val cellWSp = fontSize.sp
        val measurer = rememberTextMeasurer()
        val monoStyle = TextStyle(
            fontFamily = FontFamily.Monospace,
            fontSize   = cellWSp,
            color      = Palette.defaultFg,
        )
        val probe = remember(fontSize) { measurer.measure(AnnotatedString("M"), monoStyle) }
        val cellW: Dp = with(density) { probe.size.width.toDp() }
        val cellH: Dp = with(density) { probe.size.height.toDp() }

        val cols = max(1, (maxWidth  / cellW).toInt())
        val rows = max(1, (maxHeight / cellH).toInt())

        LaunchedEffect(cols, rows) {
            TerminalState.resize(rows, cols)
            if (!hasReady && cols > 1 && rows > 1) {
                hasReady = true
                onReady?.invoke()
            } else if (hasReady) {
                // Signal Oberon side to repaint at the new size —
                // Ctrl-L (12) is bound to "clear status" in oed and
                // handleKey unconditionally Refreshes after.
                OberonBridge.ocDispatchKey(12)
            }
        }

        Column {
            for (r in displayed.indices) {
                RowView(
                    row = displayed[r],
                    isCursorRow = cursorVisible && r == TerminalState.curRow,
                    curCol = TerminalState.curCol,
                    cellW = cellW,
                    cellH = cellH,
                    style = monoStyle,
                )
            }
        }
    }
}

@Composable
private fun RowView(
    row: Array<Cell>,
    isCursorRow: Boolean,
    curCol: Int,
    cellW: Dp,
    cellH: Dp,
    style: TextStyle,
) {
    val text = buildAnnotatedString {
        for (cell in row) {
            val cellFg = Palette.color(cell.fg) ?: Palette.defaultFg
            val cellBg = Palette.color(cell.bg) ?: Palette.defaultBg
            val (fg, bg) =
                if ((cell.attr and CellAttr.REVERSE) != 0) cellBg to cellFg
                else cellFg to cellBg
            val weight = if ((cell.attr and CellAttr.BOLD)      != 0) FontWeight.Bold   else FontWeight.Normal
            val slant  = if ((cell.attr and CellAttr.ITALIC)    != 0) FontStyle.Italic  else FontStyle.Normal
            val decor  = if ((cell.attr and CellAttr.UNDERLINE) != 0) TextDecoration.Underline else TextDecoration.None
            withStyle(
                SpanStyle(
                    color           = fg,
                    background      = bg,
                    fontWeight      = weight,
                    fontStyle       = slant,
                    textDecoration  = decor,
                    fontFamily      = FontFamily.Monospace,
                )
            ) { append(cell.ch.toString()) }
        }
    }
    Box(Modifier.height(cellH)) {
        BasicText(
            text     = text,
            style    = style,
            maxLines = 1,
            overflow = TextOverflow.Clip,
        )
        if (isCursorRow) {
            Box(
                Modifier
                    .offset(x = cellW * curCol)
                    .size(width = cellW, height = cellH)
                    .drawWithContent {
                        // Solid cursor block; BlendMode.Difference
                        // inverts the glyph underneath so the
                        // character reads through.
                        drawContent()
                        drawRect(
                            brush = SolidColor(Palette.defaultFg),
                            blendMode = BlendMode.Difference,
                        )
                    }
            )
        }
    }
}
